File: bin/hawq-config.pl
#!/usr/bin/perl
# Input:
#   $ARGV[0] property.name 
#   $ARGV[1] property.value

use v5.10;
use strict;
use warnings;
use XML::LibXML;

my $configFile = '/usr/local/hawq/etc/hawq-site.xml';
my $dom = XML::LibXML->load_xml(location => $configFile);

my $name= $ARGV[0];
my $value = $ARGV[1];

# Add property
if (!$dom->exists("//name[text()='$name']/ancestor::property")) {
	if ($value) {
		say $name, ": = ", $value;
	
		my $nameElem = $dom->createElement('name');
		$nameElem->appendText("$name");
		my $valueElem = $dom->createElement('value');
		$valueElem->appendText("$value");
	
		my $propElem = $dom->createElement('property');
		$propElem->appendChild($nameElem);
		$propElem->appendChild($valueElem);
	
		my $configElem = $dom->documentElement;
		$configElem->appendChild($propElem);
	}
} else {
	# Modify property
	foreach my $propElem ($dom->findnodes("//name[text()='$name']/ancestor::property")) {
		my $valueElem = $propElem->getChildrenByTagName('value')->get_node(0);
		if (!$value) {
			say $name, ": ", $valueElem->textContent;
		} else {
			say $name, ": ", $valueElem->textContent, "  =>  ", $value;
			$valueElem->firstChild->setData($value);
		}
	}
}


# Output file
$dom->toFile($configFile, 2);

File: bin/hawq-config.py
#!/usr/bin/env python

import sys
import xml.etree.ElementTree as et
import os.path

config_files = ['/usr/local/hawq/etc/hawq-site.xml', '/usr/local/hawq/etc/magma-site.xml']

if __name__ == '__main__':
    assert (len(sys.argv) <= 3)

    for config_file in config_files:
        if not os.path.isfile(config_file):
            continue
        tree = et.parse(config_file)
        root = tree.getroot()
        if len(sys.argv) == 2:
            name = sys.argv[1]
            value_node = root.find(".//property[name='{}']/./value".format(name))
            print('undefined' if value_node is None else value_node.text)
            exit

        if len(sys.argv) == 3:
            name = sys.argv[1]
            value = sys.argv[2]
            value_node = root.find(".//property[name='{}']/./value".format(name))
            if value_node is None:
                property_node = et.SubElement(root, 'property')
                name_node = et.SubElement(property_node, 'name')
                value_node = et.SubElement(property_node, 'value')
                name_node.text = name
                value_node.text = value
            else:
                value_node.text = value
            tree.write(config_file)
